editor/assets/default_renderpipeline/builtin-pipeline.ts

```diff
@@ -137,1295 +123,1656 @@ function setupPipelineConfigs(
+                ppl, camera, pplConfigs.mobileMaxSpotLightShadowMaps);
+        }
 
-    cameraConfigs.radianceFormat = cameraConfigs.enableHDR
-        ? gfx.Format.RGBA16F : gfx.Format.RGBA8;
-}
+        this._tryAddReflectionProbePasses(ppl, cameraConfigs, id, mainLight, camera.scene);
 
-if (rendering) {
+        if (cameraConfigs.remainingPasses > 0 || cameraConfigs.enableShadingScale) {
+            context.colorName = cameraConfigs.enableShadingScale
+                ? `ScaledRadiance0_${id}`
+                : `Radiance0_${id}`;
+            context.depthStencilName = cameraConfigs.enableShadingScale
+                ? `ScaledSceneDepth_${id}`
+                : `SceneDepth_${id}`;
+        } else {
+            context.colorName = cameraConfigs.colorName;
+            context.depthStencilName = cameraConfigs.depthStencilName;
+        }
 
-    const { QueueHint, SceneFlags, ResourceFlags, ResourceResidency } = rendering;
+        const pass = this._addForwardRadiancePasses(
+            ppl, pplConfigs, cameraConfigs, id, camera,
+            cameraConfigs.width, cameraConfigs.height, mainLight,
+            context.colorName, context.depthStencilName,
+            !cameraConfigs.enableMSAA,
+            cameraConfigs.enableStoreSceneDepth ? StoreOp.STORE : StoreOp.DISCARD);
 
-    class ForwardLighting {
-        // Active lights
-        private readonly lights: renderer.scene.Light[] = [];
-        // Active spot lights with shadows (Mutually exclusive with `lights`)
-        private readonly shadowEnabledSpotLights: renderer.scene.SpotLight[] = [];
+        if (!cameraConfigs.enableStoreSceneDepth) {
+            context.depthStencilName = '';
+        }
 
-        // Internal cached resources
-        private readonly _sphere = Sphere.create(0, 0, 0, 1);
-        private readonly _boundingBox = new AABB();
-        private readonly _rangedDirLightBoundingBox = new AABB(0.0, 0.0, 0.0, 0.5, 0.5, 0.5);
+        if (cameraConfigs.remainingPasses === 0 && cameraConfigs.enableShadingScale) {
+            return addCopyToScreenPass(ppl, pplConfigs, cameraConfigs, context.colorName);
+        } else {
+            return pass;
+        }
+    }
+    private _addCascadedShadowMapPass(
+        ppl: rendering.BasicPipeline,
+        pplConfigs: Readonly<PipelineConfigs>,
+        id: number,
+        light: renderer.scene.DirectionalLight,
+        camera: renderer.scene.Camera,
+    ): void {
+        const QueueHint = rendering.QueueHint;
+        const SceneFlags = rendering.SceneFlags;
+        // ----------------------------------------------------------------
+        // Dynamic states
+        // ----------------------------------------------------------------
+        const width = ppl.pipelineSceneData.shadows.size.x;
+        const height = ppl.pipelineSceneData.shadows.size.y;
+        this._viewport.left = 0;
+        this._viewport.top = 0;
+        this._viewport.width = width;
+        this._viewport.height = height;
```

Comment:
This code will contribute more code size than
```ts
const shadowSize = ppl.pipelineSceneData.shadows.size;
const width = shadowSize.x;
const height = shadowSize.y;

const viewport = this._viewport;
viewport.left = viewport.top = 0;
viewport.width = width;
viewport.height = height;
```

Reply:
fixed